🐛 BUG: No error handling for failed deploy due to binding with the name 'assets'.

### Which Cloudflare product(s) does this pertain to?

Pages

### What version(s) of the tool(s) are you using?

Wrangler 3.64.0

### What version of Node are you using?

20.10.0

### What operating system and version are you using?

Windows 11

### Describe the Bug

### Observed behavior
When deploying a Hono application in Wrangler, if a binding uses the name 'ASSETS' the application will fail to deploy.

### Expected behavior
The application failing to deploy is expected as assets is a reserved keyword, however no error message is provided to explain that this is the cause of the failed deployment. An error message that explains that a binding with the name assets exists and that the name needs to be changed is the expected behavior.

### Steps to reproduce
Please provide the following:
- Create a Hono application
- Add a binding with the name 'ASSETS'
- Attempt deploying the application

OK I managed to recreate this, I think. I don't think this has anything to do with Hono or R2. I think that any Pages project that is configured with a wrangler.toml that has an `ASSETS` binding (of any kind) will error with an output that looks like:

```
pnpm wrangler pages deploy 
🌍  Uploading... (1/1)

✨ Success! Uploaded 0 files (1 already uploaded) (0.16 sec)

✨ Compiled Worker successfully
✨ Uploading Worker bundle
🌎 Deploying...

✘ [ERROR] Deployment failed!

  Failed to publish your Function. Got error: Unknown internal error occurred.
  ```

Indeed I can recreate this with the following files:

_worker.js
```js
export default {
  fetch(_req, env) {
    return new Response(env.ASSETS);
  },
};
```

package.json
```json
{
  "name": "test"
}
```

wrangler.toml
```toml
name = "test"
pages_build_output_dir = "."
[vars]
ASSETS = "test"
```

Then running 

```
pnpx wrangler pages deploy 
✔ Select an account › Pete Bacon Darwin Account
🌏  Uploading... (2/2)

✨ Success! Uploaded 1 files (1 already uploaded) (1.22 sec)

✨ Compiled Worker successfully
✨ Uploading Worker bundle
🌎 Deploying...

✘ [ERROR] Deployment failed!

  Failed to publish your Function. Got error: Unknown internal error occurred.
```

--- a/.changeset/eleven-cobras-repeat.md
+++ b/.changeset/eleven-cobras-repeat.md
@@ -1,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+fix: throw a better error if there is an "ASSETS" user binding in a Pages projects

--- a/packages/wrangler/src/__tests__/configuration.pages.test.ts
+++ b/packages/wrangler/src/__tests__/configuration.pages.test.ts
@@ -767,5 +767,47 @@ describe("normalizeAndValidateConfig()", () => {
 				);
 			});
 		});
+
+		it("should error if there is a user binding named ASSETS at the top-level", () => {
+			const { diagnostics } = normalizeAndValidateConfig(
+				{
+					...pagesRawConfig,
+					kv_namespaces: [
+						{
+							binding: "ASSETS",
+							id: "1234",
+						},
+					],
+				},
+				undefined,
+				{ env: undefined }
+			);
+
+			expect(diagnostics.errors).toEqual([
+				"The name 'ASSETS' is reserved in Pages projects. Please use a different name for your KV Namespaces binding.",
+			]);
+		});
+
+		it("should error if there is a user binding named ASSETS in a named environment", () => {
+			const { diagnostics } = normalizeAndValidateConfig(
+				{
+					...pagesRawConfig,
+					env: {
+						preview: {
+							...pagesRawConfig.env?.preview,
+							vars: { ASSETS: "test_value" },
+						},
+					},
+				},
+				undefined,
+				{
+					env: "preview",
+				}
+			);
+
+			expect(diagnostics.errors).toEqual([
+				"The name 'ASSETS' is reserved in Pages projects. Please use a different name for your Vars binding.",
+			]);
+		});
 	});
 });

--- a/packages/wrangler/src/config/validation.ts
+++ b/packages/wrangler/src/config/validation.ts
@@ -2773,6 +2773,12 @@ const validateBindingsHaveUniqueNames = (
 				bindingsGroupedByName[bindingName] = [];
 			}
 
+			if (bindingName === "ASSETS" && isPagesConfig(config)) {
+				diagnostics.errors.push(
+					`The name 'ASSETS' is reserved in Pages projects. Please use a different name for your ${bindingType} binding.`
+				);
+			}
+
 			bindingsGroupedByName[bindingName].push(bindingType);
 		}
 	}

--- a/packages/wrangler/src/config/validation.ts
+++ b/packages/wrangler/src/config/validation.ts
@@ -2773,7 +2773,7 @@ const validateBindingsHaveUniqueNames = (
 				bindingsGroupedByName[bindingName] = [];
 			}
 
-			if (bindingName.toUpperCase() === "ASSETS" && isPagesConfig(config)) {
+			if (bindingName === "ASSETS" && isPagesConfig(config)) {
 				diagnostics.errors.push(
 					`The name 'ASSETS' is reserved in Pages projects. Please use a different name for your ${bindingType} binding.`
 				);

--- a/packages/wrangler/src/config/validation.ts
+++ b/packages/wrangler/src/config/validation.ts
@@ -2773,7 +2773,7 @@ const validateBindingsHaveUniqueNames = (
 				bindingsGroupedByName[bindingName] = [];
 			}
 
-			if (bindingName === "ASSETS" && isPagesConfig(config)) {
+			if (bindingName.toUpperCase() === "ASSETS" && isPagesConfig(config)) {
 				diagnostics.errors.push(
 					`The name 'ASSETS' is reserved in Pages projects. Please use a different name for your ${bindingType} binding.`
 				);

